### PR TITLE
fix: _JSXStyle is not defined

### LINKED ISSUE #695: I got _JSXStyle is not defined with 3.4.0
👋 friend. Welcome to styled-jsx and thanks for contributing!

⚠️ IMPORTANT ⚠️

If you need help or have a question about styled-jsx please ask it on Spectrum https://spectrum.chat/styled-jsx or join our Slack community at https://zeit.chat and ask it in the `#next` channel.

Before you open a new issue please take a look at our [**Frequent Asked Questions**](https://github.com/zeit/styled-jsx#faq) and [**open issues**](https://github.com/zeit/styled-jsx/issues).

Remember, often time asking in chat or looking at the FAQ/issues can be *faster than waiting for us to reply to a new issue**.

If you are here to report a bug or request a feature please remove this introductory section and fill out the information below!

---

#### What is the current behavior?
when I use <style jsx> in MyComponent I got _JSXStyle is not defined
#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

1. webpack 4.42.0
2. mac 
3. 

#### Environment (include versions)

* OS:
* Browser:
* styled-jsx (version): 3.4.0

#### Did this work in previous versions?
yes

### LINKED ISSUE #696: import _JSXStyle2 from "styled-jsx/style"; when my components's props use destructuring
👋 friend. Welcome to styled-jsx and thanks for contributing!

⚠️ IMPORTANT ⚠️

If you need help or have a question about styled-jsx please ask it on Spectrum https://spectrum.chat/styled-jsx or join our Slack community at https://zeit.chat and ask it in the `#next` channel.

Before you open a new issue please take a look at our [**Frequent Asked Questions**](https://github.com/zeit/styled-jsx#faq) and [**open issues**](https://github.com/zeit/styled-jsx/issues).

Remember, often time asking in chat or looking at the FAQ/issues can be *faster than waiting for us to reply to a new issue**.

If you are here to report a bug or request a feature please remove this introductory section and fill out the information below!

---

#### What is the current behavior?

when i use destructuring in component, i will got an error about _JSXStyle because of `import _JSXStyle2 from "styled-jsx/style";` not `import _JSXStyle from "styled-jsx/style";`

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

1. webpack 4.32.2
2. @babel/core 7.4.5

#### What is the expected behavior?
normal working
#### Environment (include versions)

* OS: 
* Browser:
* styled-jsx (version): 3.4.0 and 3.4.1；both happen

### LINKED ISSUE #711: Uncaught ReferenceError: _JSXStyle is not defined | >v3.3.2 | CRA + react-app-rewired + customize-cra + Production Build
#### Do you want to request a *feature* or report a *bug*?
Bug

#### What is the current behavior?
- Using the latest `styled-jsx` version, or any version newer than `v3.3.2` will produce a fatal error in the production build (when opened in the browser): `_JSXStyle is not defined`.
- Dependencies: create-rect-app, react-app-rewired, customize-cra.
- During development everything works as expected without any error.

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

1. Create a new app with `create-react-app`.
2. Install `react-app-rewired` and `customize-cra`
3. Configure `config-override.js`:
``` 
const { override, useBabelRc } = require("customize-cra");

module.exports = override(
  useBabelRc("./.babelrc"),
);
```
4. Configure `.babelrc` (same behavior when using the `addBabelPlugin` feature of `customize-cra`):
```
{
  "presets": ["babel-preset-react-app"],
  "plugins": [
    [
      "styled-jsx/babel",
      {
        "plugins": [
          [
            "@styled-jsx/plugin-sass",
            {
              "sassOptions": {
                "includePaths": ["./src/styles"]
              }
            }
          ]
        ]
      }
    ]
  ]
}
```

Same error happens for this `.babelrc` config:
```
{
  "presets": ["babel-preset-react-app"],
  "plugins": ["styled-jsx/babel"]
}
```

When there is no `"plugins": ["styled-jsx/babel"]` in `.babelrc` it works as expected, but then it's not possible to add plugins.

5. `yarn run build` (`react-app-rewired build`) and `serve -s build` - or deploy to bucket/server
6. Fatal error happens: `Uncaught ReferenceError: _JSXStyle is not defined`

#### What is the expected behavior?
`styled-jsx` should work in production the same as in development.

#### Environment (include versions)

* OS: Any
* Browser: Any
* styled-jsx (version): >3.3.2

#### Did this work in previous versions?
It works with v3.3.2 but not in newer versions.

Could it be an issue with this? https://github.com/vercel/styled-jsx/pull/684

#### Fix for now
- Either do `import _JSXStyle from "styled-jsx/style";` for each file.
- Or hard-code `3.3.2` in `package.json`

### PR DESCRIPTION
**Pull request by morrisallison**
### Description

`addDefault` only accepts a `nameHint` and doesn't guarantee the name of the resulting default import. The identifier returned by `addDefault` should be referenced.

This change renames the default import hint to `_JSXStyleImport`, and adds a variable declaration after the import to assign the default import to `_JSXStyle`.

As a result, instead of this being output in some cases:

```js
import _JSXStyle2 from 'styled-jsx/style';
```

This will be output:

```js
import _JSXStyleImport2 from 'styled-jsx/style';
var _JSXStyle = _JSXStyleImport2;
```

Admittedly, a better solution would be to use a named export like in #690, but this is quick fix that should unblock those who are having issues with later releases.

### Related

Fixes #695
Fixes #696
Fixes #711

### CODE REVIEW

File: test/snapshots/external.js.md [hunk 1]
```diff
@@ -78,9 +80,11 @@ Generated by [AVA](https://ava.li).
     ␊
     function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }␊
     ␊
+    var _JSXStyle = _style.default;␊
+    ␊
     function Test() {␊
       return <div>␊
-            <_style.default id={_App.default.__hash}>{_App.default}</_style.default>␊
+            <_JSXStyle id={_App.default.__hash}>{_App.default}</_JSXStyle>␊
```
Comment (morrisallison) [inline test/snapshots/external.js.md:87]: This pretty much explains what's going on.

**Comment by giuseppeg** [comment]
Hi @morrisallison thank you for taking the time to look into this. 

It seems that the problem right now is that `_JSXStyle` might be already exist in the module and `addDefault` generates an identifier with a different name. Adding a mapping to `_JSXStyleImport` is somehow similar to renaming `_JSXStyle` and hope that it will be unique.

When @jaydenseric introduced this change I anticipated this issue: https://github.com/vercel/styled-jsx/pull/684#discussion_r556031015

If we want to fix this, we will have to generate the identifier (the name of the import specifier) upfront when we enter the `Program` and pass it to the helpers that add the `<_JSXStyle>` elements like I said in the comment.

**Comment by morrisallison** [comment]
> Hi @morrisallison thank you for taking the time to look into this.
> 
> It seems that the problem right now is that `_JSXStyle` might already exist in the module and `addDefault` generates an identifier with a different name. Adding a mapping to `_JSXStyleImport` is somehow similar to renaming `_JSXStyle` and hoping that it will be unique.
> 
> When @jaydenseric introduced this change I anticipated this issue: [#684 (comment)](https://github.com/vercel/styled-jsx/pull/684#discussion_r556031015)
> 
> If we want to fix this, we will have to generate the identifier (the name of the import specifier) upfront when we enter the `Program` and pass it to the helpers that add the `<_JSXStyle>` elements like I said in the comment.
> 
> Please let me know if I misunderstood the issue that you are trying to fix.

Yup, that's accurate 👍🏽  However, I didn't want to do a lot of refactoring to reference the identifier everywhere. This change is intended to be quick (maybe dirty) fix to unblock. A proper fix will be more involved like you described.

**Comment by giuseppeg** [comment]
I'd rather fix this properly. We need to create the import identifier when we `enter` the `Program`, put it in `state` and use its `name` property (`id.node.name`) instead of `STYLE_COMPONENT`:

https://github.com/vercel/styled-jsx/blob/master/src/_utils.js#L383
https://github.com/vercel/styled-jsx/blob/master/src/_utils.js#L261
https://github.com/vercel/styled-jsx/blob/master/src/_utils.js#L624

**Comment by morrisallison** [comment]
Closing, because I don't have the bandwidth to work on further changes, but I'm glad this could help identify the problem and solution.

**Pull request closed without merging**